test/clienttester.go

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2022 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package test
+
+import (
+	"bytes"
+	"io"
+	"net"
+	"testing"
+
+	"golang.org/x/net/http2"
+)
+
+var (
+	clientPreface = []byte(http2.ClientPreface)
+)
+
+func newClientTester(t *testing.T, conn net.Conn) *clientTester {
+	ct := &clientTester{
+		t:    t,
+		conn: conn,
+	}
+	ct.fr = http2.NewFramer(conn, conn)
+	return ct
+}
+
+type clientTester struct {
+	t    *testing.T
+	conn net.Conn
+	fr   *http2.Framer
+}
+
+// greet() performs the necessary steps for http2 connection establishment on
+// the server side.
+func (ct *clientTester) greet() {
+	ct.wantClientPreface()
+	ct.wantSettingsFrame()
```

Comment:
What if the client doesn't send a settings frame?  I don't think this is a requirement, but it is something our client might always do.  Can this be handled as optional in the loop below?

Reply:
"That is, the connection preface starts with the string PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n). This sequence MUST be followed by a [SETTINGS](https://httpwg.org/specs/rfc7540.html#SETTINGS) frame ([Section 6.5](https://httpwg.org/specs/rfc7540.html#SETTINGS)), which MAY be empty." - this is what the HTTP/2 spec test says? This is also specifically testing the gRPC specific http2_client, (this is a testing util which lives in our own library). The HTTP2 client sends a settings frame in the constructor: https://github.com/grpc/grpc-go/blob/master/internal/transport/http2_client.go#L397.

Reply:
Nevermind! :D